src/Components/Components/src/RenderTree/AttributeComparerForDiffing.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    internal static class AttributeComparerForDiffing
+    {
+        private static readonly ConcurrentDictionary<Type, Func<object, object, bool>> ClosureComparerCache = new();
+        private static readonly Func<object, object, bool> AlwaysFalseDelegate = (objA, objB) => false;
+
+        internal static bool IsEquivalentForDiffing(object? oldValue, object? newValue)
+        {
+            // If the values are delegate types, we'll inspect them to see if they point to closures.
```

Comment:
I notice that, no matter which code path we go down here, we first check if `Equals(old, new)` and return `true` if so (at least I think so - please let me know if not). As such it seems like most of the type checking could be skipped in most cases if this whole method was prefixed with:

```cs
if (Equals(oldValue, newValue))
{
    return true;
}
```

Then we could simplify the logic in the other comparison methods to specifically ignore the case where the values are superficially equal.

The main reason I started thinking about this is because, as far as it seems to me, this method only ever gets called for *element* attributes (not attributes when used as component parameters), so in 90% of cases they are going to be strings. In 90% of *those* cases they will be interned strings that are reference-equal. If possible it would be good to short-circuit these code paths and not do all the checks to see if we have eventcallbacks or delegates. The `Equals` guard at the top will do that for us.

Reply:
Relatedly, I think we should change all the element-specific methods that call into here to have `ElementAttribute` in their name to distinguish them from component-parameter attributes. No need to deal with that in this PR unless you particularly want to. I can do it separately but making a note for myself here!

Reply:
Oh now I realise this method is actually used for the field comparisons inside the lambda capture object. That really needs to be based on reference equality comparisons for reference types, and only use `Equals` for specific types we know are primitive and immutable.

Reply:
In other words, it should use the same conservative equality checking logic that we do when diffing component parameter values, like this: https://github.com/dotnet/aspnetcore/blob/0011eff5648bc555e1a48711c7c24aea34afa785/src/Components/Components/src/ChangeDetection.cs#L10

Reply:
Wow, I never realised that component parameters were compared based on this method. It might be convenient for novice users who have no idea about mutable types and equality, but this means that Blazor will always rerender a component whenever you pass it the same composite reference type.

I'm pretty sure this won't work for the data binding case, because one if the things that is captured is a reference to a CultureInfo instance, which is a reference type.

Reply:
Sorry I was unclear. I wrote that during a meeting and failed to clarify something. Reference types would need to be compared with reference equality (because the object identity could be consequential, but we don’t care if it mutated) whereas for known immutable value types we’d use the logic I linked to, and for unknown value types I’m unsure if there’s a good back-compat preserving choice without recursing into its fields. 

To be honest the complexity of this is giving me real pause about the practicality of this. What do you think?

Reply:
Well, Equals normally reverts to reference equality for reference types, unless of course Equals is overridden or IEquatable is implemented. Equals will also correctly handle structs, either through direct memory comparison or memberwise compare, which both is fine. Equals is also nicely implemented for new immutable record types, no surprises there. So from a usability point of view Equals() is great (unless I misunderstand something).

Performance-wise boxing value types isn't ideal like you mentioned. I did devise a solution which avoided boxing of value types, but I was only able to pull it off by lifting the comparison code into the expression tree. The expression tree became larger, and while boxing was avoided, it was actually measurably slower than the current code, which is why I reverted it back. It doesn't make a lot of sense to create something that is more complex, harder to read, and is noticeably slower. (It might be my skills though ;-)

Reply:
For back compat, it needs not to be overridable, as people might have overridden equality to mean something different than what we were doing before.

Reply:
I agree, but the original attribute comparison was already using `Equals()`  so not much has changed here. Only in the case of a delegate pointing to a closure, we define a new situation where we compare the fields of the closure (as if it were a struct, instead of a reference to a class), which semantically makes sense to me (and incidentally also is the way equality is handled implicitly in the EventCallback struct. EvenCallbacks are the same when the receiver and delegate are the same). I don't see this as a compatibility regression. If they would have overridden Equals on a type and bound it to the attribute immediately you would have the same behaviour as when that type was captured as a free variable by a lambda.

Reply:
Can we dig into this a bit to clarify? The kind of scenario I'm concerned about looks like this:

```razor
<button @onclick="@(() => { Console.Log(somePerson.LastModifiedDate); })">Show last modified date</button>

@code {
    Person somePerson;

    class Person : IEquatable<Person>
    {
        public int PersonId { get; set; }
        public DateTime LastModifiedDate { get; set; }

        // Equality determined by PersonId
        public override bool Equals(object obj) => Equals(obj as Person);
        public bool Equals(Person other) => other?.PersonId == PersonId;
    }
}
```

In this case, if the `somePerson` value changes to a different `Person` instance that has the same `PersonId` but a different `LastModifiedDate`, when we check the two delegates we'd walk through the fields and compare those two `Person` instances and say they are the same and hence don't update the event handler. Clicking the button would show the modified date for the old instance instead of the new one, which would be a breaking change from before because we'd have compared the delegate instances and said they were different.

Am I misunderstanding this? Apologies if so. I know I've raised a lot of concerns and you're providing great answers!

At the same time I want to give a caveat that on the team we're not 100% sure we are going to be OK with introducing something new based on refemit. Generally reflection and especially refemit has very hard-to-reason-about interactions with the linker. For example if we *did* use the `Equals` method on `Person` from the refemitted-code, it's possible that no other code uses that method and so the linker would have stripped it out, and the refemitted code would throw in production. Perhaps that doesn't happen if we use `ReferenceEquals` but even so we're trying our best to do new things statically via source generators or Roslyn integration in the Razor compiler. @javiercn and @pranavkm have been suggesting we might do something at compile time using something like AnalyzeDataFlow (https://stackoverflow.com/a/30334638) to generate logic that would do the comparison, and if so that would fix the linker concerns, perf concerns, and make it usable on WebAssembly too. Potentially too complicated to attempt right now but also potentially a much stronger outcome.

Reply:
Hi Steve, sorry for my delayed response, but I was enjoying a short holiday. That holiday gave me some time to think about this, and you are right. Identity comparison for reference variables makes sense, especially since C# captures variables, not values on a closure. I'll change this and see whether we still get the benefit of reduced diff sizes.

Please note that the refemit could be easily changed to a normal reflection call to get the field value (we could only do this for the WebAssembly case, and continue to use expression trees for server-side). The cache would still be beneficial because at least field enumeration would take place only once per closure type.

In general, the question is, shouldn't this be handled on a lower level? The Roslyn compiler generates the closure object a.t.m, but most functional languages would implement the closure using a "record" type. A heap-allocated type, but with member comparison semantics like the new "record" type in C#. Historically this isn't the case with the C# compiler, and it would break backwards compatibility, but for our use case, it would make perfect sense to have this kind of comparison function generated on the closure (or its base type).

The difficulty in implementing it in the razor generator is that razor translates into C# code. At the moment of C# generation, the closure type isn't generated yet. You could generate a second lambda that captures the same variables and does stuff with it, but during Razor code generation I don't have access to the closure type or the delegate in any way, so unless I'm missing something this seems a dead end.

Any more thoughts on this?